feat(prompt): multiline prompt input

shadmansaleh opened the pull request:
This pr aims to add multiline support to the prompt buffer. 

Works toward #32420 and #32661

### Proposal:
- Track how much text has been submitted already to the prompt buffer callback.
- When the user submits, send the text after the previously submitted point.
- Make the content before last submit Readonly so it's locked in place.

### Questions:
- ~~How to handle buffer changes through API ?~~ Keep track of the location through marks so edits don't effect the position.
- ~~The callback may insert or remove text in the buffer before we lock in the submit location?~~ same as above
- Should we expose this location to user somehow?

### Example:
 
https://github.com/user-attachments/assets/5b45027d-328c-46f3-a70d-6342d3fa6d84


### Notes
- [x] Figure out how submition key should be mapped (enter, shift-enter, custom?) (enter as submit, shift+enter for newline)
       Choosing shift enter for new line and enter as submit for now.
- [x] Fix undo/redo (why doesn't it work?) (undo redo with partially modifiable buffer is complicated. I guess that's why it's disabled. What to do. a ton of edge cases. Particularly since under works in branches. Also buftype can be changed at any point) (undo/redo works up to last submit)
- [x] dd on prompt line clears the prompt too (fix it) (Entering insert mode again adds back the prompt. cc works by default)
- [x] o/O in normal mode doesn't work. it shoudl create a newline
       - check nv_open in normal.c, open_line in change.c . basic editibility check won't be enough. Probably the undo redo and dd ar the same. One key issue is operators can work on multiple lines at a time. How to handle that when only part of buffer is editable. For example with O on prompt line should we all text below it one line down and put a line there at prompt?
- [x] O at prompt location moves cursor to different line. figure out why. broken after switching to position tracking with mark for some reason. (fixed) (Although this blocks all mark updates during O operation at prompt line. Should be more granular? Is there a better way to do it other than adding a global flag?)
- [x] a/i at prompt location creates a new prompt in a newline when user input to prompt is none (fixed)
- [x] enter on normal mode submits the text should we retain this behavior? (Since we are keeping Enter as submit in insert too so keeping it as is makes sense)
- [x] When buftype is set to prompt set the last submit location to end of buffer (done)
- [x] Things like d5k doesn't work (it deletes non editable region) (broken on master too) (upon insert adds the prompt line. Was trying to block the edit. But got lost in `do_peding_operator()` . I've Tried adjusting the operator range. But doesn't seem to have effect. any suggestions?
- [x] Backspace doesn't work sometimes in prompt buffer where it should (Can't seem to reproduce now, most likely fixed with other backspace issues)

closes #32420
closes #32661

shadmansaleh committed:
wip: multiline prompt buffer

justinmk commented on src/nvim/strings.c:
```diff
@@ -574,6 +574,20 @@ char *concat_str(const char *restrict str1, const char *restrict str2)
+  size_t s = strlen(sep);
+  char *dest = xmalloc(l + s + strlen(str2) + 1);
+  STRCPY(dest, str1);
+  STRCPY(dest + l, sep);
```
since `join_str` isn't joining a list (`ga_concat_strings_sep` does that already), it's probably not that useful. the normal way to do this would be like:

    a2 = concat_str(a, ",")
    a3 = concat_str(a2, b)

shadmansaleh replied:
> (ga_concat_strings_sep does that already),

Great. I was so sure there has to be equivalent of this already in the codebase. But couldn't find it. I'll move it to reuse ga_concat_strings_sep. Or the above double call as you've suggested. I was trying to reduce the number of copies we make I think.

justinmk commented:
> Track how much text has been submitted already to the prompt buffer callback.

here be dragons. does it handle deleted text? `dd` deletes text in prompt buffers.

is there a way we can avoid maintaining this extra state? tends to be fragile.

shadmansaleh commented:
> here be dragons. does it handle deleted text? dd deletes text in prompt buffers.

It does handle text deletion in current prompt. And previous text before that is basically readonly. I think I did show c/d operations in the demo.

What I'm not sure if it handles right is probably editing actions through API. I haven't done serious testing on that but I suspect there might be bunch of edge-cases there.

> is there a way we can avoid maintaining this extra state? tends to be fragile.

True. I've been thinking about it. Before this the prompt was always lastline of buffer. When we allow multilines that's no longer true. It seems to me we have to track that state of it to work. Open to ideas though.

Also it's currently not worth reviewing. I had marely put together a quick prototype. It needs more work before it can be considered.

justinmk commented:
> It seems to me we have to track that state of it to work.

The `'[` / `']` are doing that to some extent. Ideally we can just re-use those, and any cases that are missed, can be delegated to that instead of managing it specifically for prompt buffers.

shadmansaleh commented:
> > It seems to me we have to track that state of it to work.
> 
> The `'[` / `']` are doing that to some extent. Ideally we can just re-use those, and any cases that are missed, can be delegated to that instead of managing it specifically for prompt buffers.

I'm currently not aware of what they do. Will  look into that. If there's already something keeping track of things like that, it's definitely better to use that instead of adding a new one.

shadmansaleh commented on src/nvim/edit.c:
```diff
@@ -1067,7 +1067,7 @@ static int insert_handle_key(InsertState *s)
       return 0;
     }
-    if (bt_prompt(curbuf)) {
+    if (mod_mask & MOD_MASK_SHIFT && bt_prompt(curbuf)) {
```
@justinmk this basically breaks current prompt buffer behavior. Since currently pressing <enter> directly submites the prompt. I don't think we should break this behavior. I can think of couple of alternatives

1. Make <enter> submit the user input and shift+<enter> enters a new line in prompt buffer. With it we have a question. Multiline paste needs to work with this too, obviously. But it won't directly work as it does like this. Since the paste won't have shift+enter. Is pasting tracked somewhere? Is there a way to differentiate paste from userinput? particularly when done with <C-r> in input mode?
2. Add a way to configure the submit button. set_prompt_submit_key("<cr>") for example. The default being <cr> when <cr> is used is submit key prompt buffer retains all current behavior with all it's flaws. If the submit key is set to some other key, then the multi line functionalities work.
3. Define e separate buftype for a multiline prompt buffer. separating it from the current prompt buffer behavior we have.

2 and 3 basically make multi-line support opt-in. I guess some combination of these can also be done. What do you think?

justinmk replied:
>  Is pasting tracked somewhere? Is there a way to differentiate paste from userinput? particularly when done with `<C-r>` in input mode?

yes, in `src/nvim/getchar.c`, see `paste_store` and `paste_repeat`. but, I think we shouldn't mess with that. 

instead, to avoid "breaking" the current prompt-buffer behavior, we can define a *default mapping* so that `<Enter>` continues to work as users expect. but plugins/users can easily support Enter-as-newline by *unmapping* that default mapping.

thus, internally, at a low-level, we use something like you have here to "submit". and we just define a default mapping for Enter (`:nnoremap <CR> <S-CR>`) in prompt-buffers.

> Add a way to configure the submit button. set_prompt_submit_key("") for example

can't that be a mapping (see above)? 

> Define e separate buftype for a multiline prompt buffer.

don't think we need this.

shadmansaleh committed:
use concat_str to build line buffer

shadmansaleh committed:
Behavior fix(1)
- A/I in editable region works as usual
- i/a/A/I outside of editable region moves cursor to end of current
  prompt
- only consider inserting prompt text when cursor is at prompt line
  (fixes a at prompt text generating a newline with prompt text again)

shadmansaleh committed:
enable o/O in prompt buffer

shadmansaleh commented:
@justinmk I'm seeing that lots of the editing stuff doesn't support the prompt buffer. For example, undo/redo is disabled. Things like range edits (like `d5k` in normal mode) can modify read-only portions of the buffer. The main issue is the read-only and writeable segments in the same buffer. Good part of neovim treats it as non-existent. Or broadly disables the feature to avoid facing it. Can you check the Notes section in the pr? And provide your opinion on those? I feel like achieving 100% regular buffer like editing experience, will be pretty hard. But we can probably get to 90% of that without way too much effort. I'm not sure how to deal with undo/redo yet, though.

Also, I looked at '[ , '] marks. They basically kind of do the same. They work by adding Pos_T variables (b_op_start,, b_op_end) to the file_buffer struct, like I'm adding the line number here. Then the mark system uses those values from the buffer to expose those positions as marks. Even if we go that route, we'll be adding an additional state. With marks, we can expose it to the user, but that takes away at least a key for the mark spot. I'm not sure if any special ones are still unused. Also, we can't directly overload '[/ '] marks. They kind of represent completely different thing from what we are trying to achieve here.

shadmansaleh committed:
when buftype is set to prompt initialize last submit to last line of buffer

shadmansaleh committed:
in prompt buffer use shift+enter to start a new line and plane enter to submit

justinmk commented:
> I feel like achieving 100% regular buffer like editing experience will be complicated. But we can probably get to 90% of that without way there. I'm not sure how to deal with undo/redo yet, though.

yeah, I don't think 100% is a goal. the idea of https://github.com/neovim/neovim/issues/32420 is primarily that the *input* (writable) section can intuitively support multiline input.

let's consider undo/redo out of scope, they aren't useful in prompt buffers afaik.

however, a major goal is that it should be easy for plugins to get the text of the *current prompt* (multiline or not) and be able to *update* the unsubmitted prompt. that avoids the need to support every editing feature, because then plugins can do fancy stuff in e.g. a floating window.

> How to handle buffer changes through API ?

plugins are free to do arbirary things to prompt-buffers. there is no defined expectation for this.

the important thing is that we maintain a meaningful region for the "writeable" (prompt) region vs the "readonly" region.

> The callback may insert or remove text in the buffer before we lock in the submit location?

the `'[` or `'.` mark tells us the last-changed part. we shouldn't be trying to calculate that any other way. if that mark is telling us the wrong thing, we need to find out why and perhaps fix it.

shadmansaleh commented:
> however, a major goal is that it should be easy for plugins to get the text of the current prompt (multiline or not) and be able to update the unsubmitted prompt

That much is already done. 

> the important thing is that we maintain a meaningful region for the "writeable" (prompt) region vs the "readonly" region.

My question basically was, do we block changes in the readonly section from the API too? Or should we allow api to make arbitrary changes to the buffer? Even on the ReadOnly section. That gives extra power to the plugins. Technically, we can't exactly block it either. As the plugin can just switch the buftype to regular text make arbitrary changes, then switch back to the prompt buffer. So maybe it's not worth blocking APIs.

> the '[ or '. mark tells us the last-changed part. we shouldn't be trying to calculate that any other way. if that mark is telling us the wrong thing, we need to find out why and perhaps fix it.

That's different. It's doing what it's supposed to do. For example, say you are editing something in insert mode. Then went to different line edited that. The marks are keeping track of that. But that's not what we need we need the start of current prompt. Modifying the prompt is not one edit actions. There can be multiple as user switches between modes or applies some operator. The '[ mark will track only the last edit action not the whole prompt edit.

> The callback may insert or remove text in the buffer before we lock in the submit location?

That question is resolved now. We track the last submit location after callback is done. So any modifications done in the callback would be tracked. What we are not tracking is edits through api. I think I should take a look at mark/extmark subsystem to see if we can leverage that to keep track of api edits.

justinmk commented:
> My question basically was, do we block changes in the readonly section from the API too?

Plugins (e.g. AI chat plugins) need to be able to show arbitrary stuff in the "readonly" section.

Why would we want to block changes to the readonly section? Our central task is to correctly set the "prompt" region, and we don't care what happens in the "readonly" region as long as the user *editing* UX isn't weird.

Using a extmark `nvim_buf_set_extmark()` may "track" changes, or at least it should in theory. Try setting an extmark on the readonly "boundary", and it should adjust as edits are made (by plugin code) to the "readonly" region.

shadmansaleh committed:
enable undo/redo in prompt buffer

shadmansaleh committed:
oldtest fix?

shadmansaleh commented:
Can Neovim detect Shift+Enter on all kinds of terminals? Or at least on most of them? I primarily use kitty. But also tried out on konsole (kde terminal) there it doesn't seem to detect shifted <CR>.

Now I feel like assigning <CR> for submit and <S-CR> as insert line is better. Because,
1. Keeps backward compatibility. CR just submits as before
2. <S-CR> is always available for entering multiple lines on all prompt buffers. Plugins don't need to add explicit support.
    if they don't want multiline text they can just map <S-CR> to <NOP>  `inoremap <S-CR> <NOP>`. It's also pretty simple to just split the input before processing in the callback
3. If in some terminal, neovim can't detect <S-CR> prompt buffer still works. Although multiline won't work. At least, single-line inputs will be working. Technically, multiline can still be done with o/O in normal mode even in that case. On the other hand, if we map <CR> to <S-CR> and treat that as submit prompt buffers will be completely broken. As submitting won't work without <S-CR> being detected.
4. If some plugin want's <CR> to add newline and <S-CR> to be submit. They can just remap the keys.
     inoremap <CR> <S-CR>
     inoremap <S-CR> <CR>

Particularly, 2 and 3 are clear wins compared to the other approach, and we don't lose anything doing it this way.

shadmansaleh committed:
keep track of next prompt with marks instead of linenr_T

shadmansaleh committed:
apply formatter

shadmansaleh committed:
keep the last submit mark even if line is deleted

shadmansaleh committed:
fix memory leak on undo clear + clear undo history even if prompt callback is not set

shadmansaleh committed:
keep the uneffected asserts

shadmansaleh committed:
add tests

shadmansaleh committed:
fix undo being able to undo text inserted by prompt callback

shadmansaleh committed:
more tests on undo and o/O

shadmansaleh committed:
fix lint

shadmansaleh committed:
update docs

zeertzjq commented on test/old/testdir/test_prompt_buffer.vim:
```diff
@@ -166,11 +166,8 @@ func Test_prompt_buffer_edit()
   call assert_beeps('normal! ~')
-  call assert_beeps('normal! o')
-  call assert_beeps('normal! O')
   call assert_beeps('normal! p')
   call assert_beeps('normal! P')
```
```suggestion
  " call assert_beeps('normal! o')
  " call assert_beeps('normal! O')
  call assert_beeps('normal! p')
  call assert_beeps('normal! P')
  " call assert_beeps('normal! u')
```

shadmansaleh replied:
Ohh right, those, cleaning them.

zeertzjq commented on src/nvim/ops.c:
```diff
@@ -1391,6 +1391,7 @@ int insert_reg(int regname, yankreg_T *reg, bool literally_arg)
           // Insert a newline between lines and after last line if
           // y_type is kMTLineWise.
           if (reg->y_type == kMTLineWise || i < reg->y_size - 1) {
+            // likely the place to insert a shift+NL in prompt buffers figure out how to insert it (shadman)
```
?

shadmansaleh replied:
This is basically how inserting register is supposed to work in insert mode with `<C-r>`. Currently, it just raw inserts the reg. As a result, if the reg has multiple lines, it just submits the prompt. I think it makes sense to paste the reg completely like `p` is normal mode. For that, we need to insert `<S-CR>` in place of `<CR>`  here on the prompt buffers. Any idea how to do so? I had asked on the dev channel as well as justin on element but didn't get an answer. I thought you or bfredl would be the best person to ask on these stuff but didn't yet got around to it. I had left the comment as a marker so I can find the location later :]

Also another thing is do we want `<C-r>` to be raw input? In that case keeping the current behavior is fine.

zeertzjq replied:
I mean, why is the comment here?

shadmansaleh replied:
It was a temporary comment. To mark where change needs to be made. So I can find it later without having to read most of the logic there again. I'm going to remove it before merge.

shadmansaleh committed:
allow pasting at prompt text in last prompt line.

In this case we paste the text at first editable location.
Convinience feature. Otherwise it does not make sense that
you need to add a space before pasting just to get a editable postion

shadmansaleh committed:
fix oldtests. also commentout the uneccesary asserts

shadmansaleh committed:
Fix O on at prompt line.

Since O at prompt line moves the replaces the old prompt line with a new
one the prompt submit mark needs to stay in place.

shadmansaleh committed:
more robust fix for O at current prompt

shadmansaleh committed:
fix backspace borken behavior

shadmansaleh committed:
fix lint

shadmansaleh committed:
fix prompt getting added to wrong line also better fix for backspace issue

shadmansaleh committed:
fix flacky undo tests

shadmansaleh committed:
more tests

shadmansaleh committed:
add promptline when prompt line has been removed

shadmansaleh committed:
cleanup

shadmansaleh committed:
test cleanup

justinmk commented on src/nvim/change.c:
```diff
@@ -1746,7 +1746,23 @@ bool open_line(int dir, int flags, int second_line_indent, bool *did_do_comment)
+  char *prompt_moved = NULL;
   if (dir == BACKWARD) {
+    if (bt_prompt(curbuf)
+        && curwin->w_cursor.lnum == curbuf->b_prompt_submitted.mark.lnum) {
```
line-length limit is 100 (just mentioning this for future reference)

justinmk commented on src/nvim/mark.c:
```diff
@@ -1223,6 +1223,11 @@ void mark_adjust_buf(buf_T *buf, linenr_T line1, linenr_T line2, linenr_T amount
 
+    // on prompt buffer adjust the last prompt start location mark
+    if (bt_prompt(curbuf)) {
+      ONE_ADJUST_NODEL(&(buf->b_prompt_submitted.mark.lnum));
```
Nice, this looks like the right approach

justinmk commented on src/nvim/optionstr.c:
```diff
@@ -695,6 +696,11 @@ const char *did_set_buftype(optset_T *args)
       || opt_strings_flags(buf->b_p_bt, opt_bt_values, NULL, false) != OK) {
     return e_invarg;
   }
+  // when buftype is set to prompt set the last prompt submit pos to lastline
```
hard to understand this sentence. is this a correct rewording?

```suggestion
  // buftype=prompt: set the prompt start position to lastline.
```

justinmk reviewed (approved):
This looks great!

justinmk commented:
> It only matters for `<C-r>` in insert mode. If we go with the `<C-r>` rework you proposed yesterday that also becomes irrelevant

It becomes less urgent, but is still relevant for `<c-r>=@x` and probably some other situations.

Not a blocker for now, we can see if it matters in practice.

justinmk commented:
> Should we expose this location to user somehow?

We definitely want some way for plugins to easily ask "what is the current prompt text" without mucking around with getline(), marks, etc. Is that part of this PR?, if not can be a followup.

shadmansaleh committed:
address review

shadmansaleh commented on src/nvim/buffer_defs.h:
```diff
@@ -699,6 +699,7 @@ struct file_buffer {
   Callback b_prompt_interrupt;  // set by prompt_setinterrupt()
   int b_prompt_insert;          // value for restart_edit when entering
                                 // a prompt buffer window.
+  fmark_T b_prompt_submitted;   // Start of current prompt insert
```
updated

shadmansaleh commented on src/nvim/change.c:
```diff
@@ -1746,7 +1746,23 @@ bool open_line(int dir, int flags, int second_line_indent, bool *did_do_comment)
+  char *prompt_moved = NULL;
   if (dir == BACKWARD) {
+    if (bt_prompt(curbuf)
+        && curwin->w_cursor.lnum == curbuf->b_prompt_submitted.mark.lnum) {
```
updated

shadmansaleh commented:
I've fixed the concerns you had in review.

> Not a blocker for now, we can see if it matters in practice.

Great. Then we can keep it like this for now. We can see it in the master. Until 0.12 gets released, the behavior isn't fixed either way.

> After the few minor comments above, this is probably ready to merge imo, though I might take a closer look at the tests later.

Do take a look as you get time, and let me know if any change is needed or if you need any clarification. Also, do I need to get commitlint to pass? Doing squash merge should be enough, right?  If you want I can do a rebase and squash all the commits.

> We definitely want some way for plugins to easily ask "what is the current prompt text" without mucking around with getline(), marks, etc. Is that part of this PR?, if not can be a followup.

It's not there for now. Currently plugins can only get the prompt text when user submits. Though adding an api for that is not complex ( we just need to agree on a api signature/name :] ) . I can do it in a followup..

justinmk commented:
> Also, do I need to get commitlint to pass?

no i'll squash

zeertzjq commented on test/old/testdir/test_prompt_buffer.vim:
```diff
@@ -166,11 +166,6 @@ func Test_prompt_buffer_edit()
   normal! i
   call assert_beeps('normal! dd')
   call assert_beeps('normal! ~')
```
```suggestion
  call assert_beeps('normal! ~')
  " Nvim: these operations are supported
  " call assert_beeps('normal! o')
  " call assert_beeps('normal! O')
  " call assert_beeps('normal! p')
  " call assert_beeps('normal! P')
  " call assert_beeps('normal! u') 
```

shadmansaleh replied:
Ok, makes sense. keeping it.

zeertzjq commented:
Document this in vim_diff.txt

shadmansaleh committed:
address review (zeertzjq)

shadmansaleh commented:
> is this tracked in todo list / issue somewhere ?

I've added it to my gsoc tracking discussion. And pr open at #34491. Will finalize it after this gets merged.

shadmansaleh committed:
expose prompt location as ':' mark

shadmansaleh committed:
add tests for mark

shadmansaleh committed:
fix lint

shadmansaleh committed:
mention : mark is only available in prompt buffer in docs

justinmk commented:
> I don't think that's a good idea. visually we won't be able to tell where actually the prompt starts like you mentioned.
> 
> >

seems to be the current behavior when i use `<s-enter>`

justinmk commented on test/functional/legacy/prompt_buffer_spec.lua:
```diff
@@ -245,4 +246,280 @@ describe('prompt buffer', function()
+    ]])
+  end)
+
+  it('can paste multiline text', function()
```
also need a test for `nvim_paste()` explicitly, since that is a rather different codepath.

justinmk commented:
> No. What terminal are you testing on?

I tested with `nvim_input('ifoo<s-enter>')` (terminal-independent)

mfussenegger commented:
What's a bit odd with the changes here is that now `i`/`a` allow to edit the prompt text and also no longer jump to the end of the prompt text depending on where the cursor is. E.g. If I'm a line above `a` will jump to the end, but if the cursor is within the prompt text, it won't. Is that intentional?
